plugins/out_azure_kusto/azure_kusto_ingest.c

```diff
@@ -0,0 +1,427 @@
+        flb_upstream_conn_release(u_conn);
+        return NULL;
+    }
+    uri = flb_sds_printf(&uri, "%s/%s.multijson?%s", blob_uri, blob_id, blob_sas);
+
+    flb_plg_debug(ctx->ins, "uploading payload to blob uri: %s", uri);
+
+    c = flb_http_client(u_conn, FLB_HTTP_PUT, uri, payload, payload_size, NULL, 0, NULL,
+                        0);
+    if (!c) {
+        flb_plg_error(ctx->ins, "cannot create HTTP client context for blob container");
+        flb_upstream_conn_release(u_conn);
+        flb_sds_destroy(uri);
+        return NULL;
+    }
+
+    flb_http_add_header(c, "User-Agent", 10, "Fluent-Bit", 10);
+    flb_http_add_header(c, "Content-Type", 12, "application/json", 16);
+    flb_http_add_header(c, "x-ms-blob-type", 14, "BlockBlob", 9);
+    flb_http_add_header(c, "x-ms-date", 9, tmp, len);
+    flb_http_add_header(c, "x-ms-version", 12, "2019-12-12", 10);
+
+    ret = flb_http_do(c, &resp_size);
+    flb_plg_debug(ctx->ins, "kusto blob upload request http_do=%i, HTTP Status: %i", ret,
+                  c->resp.status);
+    if (ret != 0) {
+        flb_plg_error(ctx->ins, "cannot send HTTP request");
+        flb_upstream_conn_release(u_conn);
+        flb_sds_destroy(uri);
+        flb_http_client_destroy(c);
+        return NULL;
+    }
+
+    /* Validate return status and HTTP status if set */
+    if (c->resp.status != 201) {
+        if (c->resp.payload_size > 0) {
+            flb_plg_debug(ctx->ins, "Request failed and returned: \n%s", c->resp.payload);
+        }
+        else {
+            flb_plg_debug(ctx->ins, "Request failed");
+        }
+        flb_upstream_conn_release(u_conn);
+        flb_sds_destroy(uri);
+        flb_http_client_destroy(c);
+        return NULL;
+    }
+
+    /* generate full blob uri & return it */
+    full_uri = flb_sds_create_size(9 + flb_sds_len(u_node->host) + flb_sds_len(uri));
+    if (!full_uri) {
+        flb_plg_error(ctx->ins, "cannot create full blob uri");
+        flb_upstream_conn_release(u_conn);
+        flb_sds_destroy(uri);
+        flb_http_client_destroy(c);
+        return NULL;
+    }
+    full_uri = flb_sds_printf(&full_uri, "https://%s%s", u_node->host, uri);
+
+    flb_plg_debug(ctx->ins, "created blob %s", full_uri);
+
+    flb_upstream_conn_release(u_conn);
+    flb_sds_destroy(uri);
+    flb_http_client_destroy(c);
+
+    return full_uri;
+}
+
+static flb_sds_t create_ingestion_message(struct flb_azure_kusto *ctx, flb_sds_t blob_uri,
+                                          size_t payload_size)
+{
+    flb_sds_t message;
+    char *uuid;
+    char *message_b64;
+    size_t b64_len;
+
+    uuid = generate_uuid();
+    if (!uuid) {
+        flb_plg_error(ctx->ins, "error generating unique ingestion UUID");
+        return NULL;
+    }
+
+    message = flb_sds_create(NULL);
+    if (!message) {
+        flb_plg_error(ctx->ins, "error creating ingestion message buffer");
+        flb_free(uuid);
+        return NULL;
+    }
+    message = flb_sds_printf(
```

Comment:
`flb_sds_printf` could return `NULL` either this time or the following if the `flb_sds_increase` (`realloc`) fails so you need to store the result without overwriting the original variable to be able to release the resource if the call fails (same applies to the next call)

Reply:
I see, will do.

What about other `printf`s where I pre-allocate the string with the expected size (for instance, https://github.com/papigers/fluent-bit/blob/293cbe92fa39442ff42735cfde7fdbf0b571efeb/plugins/out_azure_kusto/azure_kusto_ingest.c#L135-L141), should these be fine?

Reply:
If you check the code for [flb_sds_printf](https://github.com/fluent/fluent-bit/blob/6af60e6d75d3196d67cb041fc51dff14a9dd4da7/src/flb_sds.c#L408) you'll notice that if it does need to adjust the size it modifies the pointer it's passed so considering that those functions don't accept an uninitialized sds (ie. an `flb_sds_t` that holds `NULL`) the only use I can come up with for the result is determining if the variable was rellocated or if the function failed.

To clarify, considering that the function automatically adjusts the variable that holds the sds pointer if it reallocates it I'd take this result purely as a status and avoid using it to update the variable. This does make me think that we should talk about the design and try to improve it in the future but I won't waste your time with my rambling.

Reply:
Changed all calls to use `flb_sds_snprintf`, I omitted checks when the buffer was already pre-allocated with the expected size of the string, hope that's fine.